### PR TITLE
Gjoshi/fixing build status posting flow

### PR DESCRIPTION
**Pull request by goravjoshi**
Proper handling of build status

### CODE REVIEW

**Comment by goravjoshi** [comment]
FYI, @alibk0rd below are the pending Tasks

1. Finish Unit testing of `LocalSCMListener` class
2. Add Assertion on the body of build posting URL in `BuildStatusPosterIT`. For example, line 80 will need to assert POST body as well. 
3. Manual test of an upgrade of Bitbucket Server plugin.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCM.java [hunk 1]
```diff
@@ -140,10 +137,11 @@ public BitbucketSCM(
 
     /**
      * Regenerate SCM by looking up new repo URLs etc.
+     *
      * @param oldScm old scm to copy values from
      */
     public BitbucketSCM(BitbucketSCM oldScm) {
-        this(oldScm.getId(), oldScm.getBranches(), oldScm.getCredentialsId(), oldScm.getExtensions(),
+        this(oldScm.getId(), oldScm.getBranches(), oldScm.getCredentialsId(), emptyList(),
```
Comment (alibk0rd) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCM.java:144]: Is this change necessary? What's the harm in still passing the extensions to the new instance (even if we don't have any extensions at the moment)?

File: src/test/java/it/com/atlassian/bitbucket/jenkins/internal/status/BuildStatusPosterIT.java [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+                format("/rest/api/1.0/projects/%s/repos/%s/commits/%s/builds/([a-z0-9]*)", PROJECT_KEY, repoSlug, latestCommit);
+        bitbucketProxyRule.getWireMock().stubFor(post(
+                urlPathMatching(url))
+                .willReturn(aResponse().withStatus(HttpStatus.SC_NO_CONTENT)));
+
+        jenkinsProjectHandler.runPipelineJob(wfj);
+
+        verify(postRequestedFor(urlPathMatching(url))
+                //TODO
+//                .withRequestBody(
```
Comment (dkjellin) [inline src/test/java/it/com/atlassian/bitbucket/jenkins/internal/status/BuildStatusPosterIT.java:180]: No thrilled about lots of commented out code, do we plan on addressing this TODO before merging? If not I think we should remove this block.

File: src/test/java/it/com/atlassian/bitbucket/jenkins/internal/status/BuildStatusPosterIT.java [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+                format("/rest/api/1.0/projects/%s/repos/%s/commits/%s/builds/([a-z0-9]*)%%2F([a-z0-9]*)", PROJECT_KEY, repoSlug, latestCommit);
+        bitbucketProxyRule.getWireMock().stubFor(post(
+                urlPathMatching(url))
+                .willReturn(aResponse().withStatus(HttpStatus.SC_NO_CONTENT)));
+
+        jenkinsProjectHandler.performBranchScanning(mbp);
+        jenkinsProjectHandler.runWorkflowJobForBranch(mbp, "master");
+
+        verify(postRequestedFor(urlPathMatching(url))
+                //TODO
```
Comment (dkjellin) [inline src/test/java/it/com/atlassian/bitbucket/jenkins/internal/status/BuildStatusPosterIT.java:226]: Same here, I think we should either fix the todo or remove the comment

File: src/test/java/it/com/atlassian/bitbucket/jenkins/internal/status/BuildStatusPosterIT.java [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+        String latestCommit = gitHelper.pushEmptyCommit("test message");
+        url =
+                format("/rest/api/1.0/projects/%s/repos/%s/commits/%s/builds/([a-z0-9]*)", PROJECT_KEY, repoSlug, latestCommit);
+        bitbucketProxyRule.getWireMock().stubFor(post(
+                urlPathMatching(url))
+                .willReturn(aResponse().withStatus(HttpStatus.SC_NO_CONTENT)));
+        jenkinsProjectHandler.runPipelineJob(wfj);
+
+        verify(postRequestedFor(urlPathMatching(url))
+                //TODO
```
Comment (dkjellin) [inline src/test/java/it/com/atlassian/bitbucket/jenkins/internal/status/BuildStatusPosterIT.java:279]: Here as well, I think we need to either address the todo or remove the comment

**Comment by alibk0rd** [comment]
@goravjoshi I think I've addressed all the comments and remaining tasks. Once you're happy with the changes and the builds are green, I'll merge.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketClientFactoryImpl.java [hunk 5]
```diff
@@ -31,11 +31,12 @@ public BitbucketCapabilitiesClient getCapabilityClient() {
     }
 
     @Override
-    // TODO: could bitbucketSCMRepo be null?
-    public BitbucketBuildStatusClient getBuildStatusClient(String revisionSha, @Nullable
-            BitbucketSCMRepository bitbucketSCMRepo, BitbucketCICapabilities ciCapabilities) {
+    public BitbucketBuildStatusClient getBuildStatusClient(String revisionSha,
+                                                           @Nullable BitbucketSCMRepository bitbucketSCMRepo,
```
Comment (goravjoshi) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketClientFactoryImpl.java:35]: Does this need to be nullable? Would there ever be an execution that reaches this place and repo being null? If not, we should remove nullable annotation.

Comment (alibk0rd) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketClientFactoryImpl.java:35]: Ah yes, I checked this by both mentally following the code path and debugging with a breakpoint, and I very much doubt this will ever be null, because as we discussed in-person, it makes no logical sense for the `BitbucketSCMRepository` to be null when we're posting a Bitbucket-SCM-based build status back to Bitbucket, but I wasn't sure if I could definitely prove so with 100% certainty. I'm happy to treat this as `@Nonnull` and put some safeguards in place _just in case_ it's ever null.

Comment (goravjoshi) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketClientFactoryImpl.java:35]: Yes, we should mark it as non-null. In general, we should be aggressively removing null as a value. Case in point, it wasted my time and now yours to follow null values.
